### DISCUSSION TITLE
ROS not initializing on HSR bootup. (No ros topics available - "Unable to Communicate with Master")

### DISCUSSION THREAD

**Discussion by siddharthbhurat4** [HumanSupportRobot/QandA#95] [Q&A]
ROS on the HSR does not initialize at all on starting/restarting the HSR. After power on, its status indicator LED infinitely stuck at initial white light not blinking, just solid.

Also when I do rostopic list on external computer in the hsrb_mode, it says "Unable to communicate with Master".

It does get connected to wifi network. This I checked by ping hsrc.local on the external computer and also by pinging my external computer from HSR. Both of them work, which means it is connected to wifi. But interestingly, when I login to the rout er portal and check the list of connected devices

Please can I know, how to fix this issue

**Comment by masudam** [comment]
Autostart of HSR works by starting scripts in docker, which is started by ubuntu services.
So I think you should check whether the ubuntu services and docker are working correctly.

login to HSR as administrator, and check services by command below.
```
systemctl status docker.hsrb.roscore.service
systemctl status docker.hsrb.robot.service
```

or check docker logs.
```
docker logs -f docker.hsrb.roscore.service
docker logs -f docker.hsrb.robot.service
```
When you check docker.hsrb.robot.service, try emergency button off/on to check autostart!

I think something in these give error.

**Reply by siddharthbhurat4** [reply]
I tried running these commands and these are the outputs I get - 
![image](https://github.com/HumanSupportRobot/QandA/assets/114542819/25698ab2-0494-4ec7-8e0e-d2a80d898b28)
![image](https://github.com/HumanSupportRobot/QandA/assets/114542819/64565b47-0545-4271-b58a-d6254ba5c941)
![image](https://github.com/HumanSupportRobot/QandA/assets/114542819/05f0dcc9-c221-4f67-8500-a41fbc3b2888)

Please can you guide me on how to fix this
Thank you

**Reply by masudam** [reply]
It seems that docker.hsrb.roscore.service had errors (in 1st pic) and other services couldn't launch because of that.
So we should try to check what happened in docker.hsrb.roscore.service next.
Could you show ` journalctl -xeu docker.hsrb.roscore.service ` ? This will show recent detailed logs of service.

**Reply by siddharthbhurat4** [reply]
This is the output of - `journalctl -xeu docker.hsrb.roscore.service`
![image](https://github.com/HumanSupportRobot/QandA/assets/114542819/ab97bec0-796c-4e6a-9e0f-e8c40903fcdc)

**Reply by siddharthbhurat4** [reply]
It keeps on trying and repeating this error output

**Reply by masudam** [reply]
I think 'docker run' command  executing in the service doesn't work well (Maybe the arguments are not being passed correctly) .
You can see what this service is going to do in `cat /etc/systemd/system/docker.hsrb.roscore.service`, so watch service file and check its docker images and environment files exist.

docker image:
 Does required docker image exist?  Maybe its name is 'docker.hsr.io/hsrb/robot' and we can see in `docker images`.

environment files:
 Three env files are used in service (written in docker.hsrb.roscore.service file above)
```
EnvironmentFile=-/etc/opt/tmc/robot/version
EnvironmentFile=-/etc/opt/tmc/robot/docker.hsrb
EnvironmentFile=-/etc/opt/tmc/robot/docker.hsrb.user
```
Please check they are in path. (`cat /etc/opt/tmc/robot/version` ...)

Also check if you can start the following docker manually.
`docker run --rm --name test --tty --env ROBOT_VERSION docker.hsr.io/hsrb/robot:noetic rosrun hsrb_bringup hsrb_roscore_service`

**Reply by siddharthbhurat4** [reply]
I tried checking all that you mentioned:

Docker Image - yes it does exist.
![image](https://github.com/HumanSupportRobot/QandA/assets/114542819/fb5aaf81-e1fb-44b2-9256-1cbedc1a54d9)

Environment Files - yes those are all there.
![image](https://github.com/HumanSupportRobot/QandA/assets/114542819/7356637b-4f39-4349-b11b-d1784914ceb1)

If I can run it manually - Yes, I can run it manually, but I cannot see any available topics and the robot also doesnt move as it generally does when it boots up and initialize.
![image](https://github.com/HumanSupportRobot/QandA/assets/114542819/35b6707c-9bf1-4af5-8d44-fa93d038b8f1)

**Reply by masudam** [reply]
Hmm.. so check the service file is not changed.

see /etc/systemd/system/docker.hsrb.roscore.service and compare that and correct description under
```
[Unit]
Description=HSRB Core service
After=docker.service
Requires=docker.service
 
[Service]
EnvironmentFile=-/etc/opt/tmc/robot/version
EnvironmentFile=-/etc/opt/tmc/robot/docker.hsrb
EnvironmentFile=-/etc/opt/tmc/robot/docker.hsrb.user
ExecStartPre=/usr/bin/install -d -m755 -o ros -g ros /var/run/ros
ExecStartPre=/usr/bin/install -d -m755 -o ros -g ros /var/run/ros/workspace
ExecStartPre=/bin/bash -c '. /etc/opt/tmc/robot/roshost.sh; /bin/systemctl set-environment $(get_roshost)'
ExecStartPre=-/usr/bin/docker rm -f %n
ExecStartPre=/bin/sh -c 'exec /usr/bin/docker run \
--name %n \
  --detach \
--tty \
--volume=${WORKSPACE}:${WORKSPACE}:ro \
${DOCKER_OPTS} \
${DOCKER_EXTRA_OPTS} \
${IMAGE_REPOSITORY}/hsrb/robot:${IMAGE_VERSION} \
rosrun hsrb_bringup hsrb_roscore_service'
ExecStart=/usr/bin/docker logs -f %n
ExecStop=-/usr/bin/docker stop %n
ExecStopPost=-/usr/bin/docker stop %n
SyslogIdentifier=%n
Restart=always
RestartSec=10s
Type=simple

[Install]
WantedBy=multi-user.target
```
